docs/proposals/PROPOSAL_E.md

```diff
@@ -0,0 +1,140 @@
+# Proposal E - Platform Features
+
+This proposal leverages the previously reserved `features` field in the platform to give a finer grain of control to runtimes selecting an image to run.
+
+The goal of this proposal is to define hard requirements that would prevent a binary from being run on a given hardware platform.
+The key differences between proposal A and E are that proposal E:
+
+- Requires each feature to be predefined in the image-spec, rather than reverse DNS namespaced externally managed values.
+- Requires runtimes to be aware of each feature, and reject an index entry with an unrecognized or unsupported feature.
+- Requires runtimes detect support of features directly, rather than from user injected configurations.
+
+## Links
+
+| Description | Link |
+|-------------|------|
+| Image Index Spec | <https://github.com/opencontainers/image-spec/blob/main/image-index.md> |
+
+## Modifications
+
+### Image Spec
+
+The following language is already included in the Image Index definition for the `platform.features`:
+
+> This property is RESERVED for future versions of the specification.
+
+This would replace that text with:
+
+> This OPTIONAL property specifies an array of strings, each specifying a mandatory feature that implementations SHOULD understand.
+> Feature values may be architecture specific.
+
+There would also be a table of known features or links to specifications, similar to how [platform variants](https://github.com/opencontainers/image-spec/blob/main/image-index.md#platform-variants) are defined.
+A possible source for known features is the [golang.org/x/sys/cpu](https://pkg.go.dev/golang.org/x/sys/cpu) package, where  each feature could be the string after `Has` in the struct for the given platform (converting to lowercase).
+
+An example would also be added to the Image Index:
+
+```jsonc
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.index.v1+json",
+  "manifests": [
+    {
+      "mediaType": "application/vnd.oci.image.manifest.v1+json",
+      "size": 7143,
+      "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
+      "platform": {
+        "architecture": "amd64",
+        "os": "linux"
+      }
+    },
+    {
+      "mediaType": "application/vnd.oci.image.manifest.v1+json",
+      "size": 7682,
+      "digest": "sha256:5b0bcabd1ed22e9fb1310cf6c2dec7cdef19f0ad69efa1f392e94a4333501270",
+      "platform": {
+        "architecture": "amd64",
+        "os": "linux",
+        "variant": "v2",
+        "features": [ "avx2", "bmi2" ]
+      }
+    }
+  ]
+}
+```
+
+In the above scenario, the generic `linux/amd64` entry is listed first and should be used by default.
+Runtimes that recognize and support the `avx2` AND `bmi2` features on a "v2" or newer variant would prefer the second entry as a better match.
```

Comment:
Overall I like the proproposal for its simplicity in image selection as I mentioned in previous comment.

just one question here, does it assume that the matching criteria is the more features match, the more prefered image.  How does the runtime decide when two entries have same number of features, and both matched with the node's CPU features. (I am OK first come first serve)

Reply:
Good question! Entries with features listed would want to be listed after entries without any features to avoid breaking older clients that are unaware of features and would pick the first entry. I think among multiple matching entries with features specified, picking the first may be the most flexible for implementation.

Otherwise, trying to count the number of matching features and using the first match for ties could also work. My concern with that approach is that I could foresee the fallback to fast instruction being a combination of several slower instructions, and we don't want to prioritize the latter.